raftinc/allocate.hpp

```diff
@@ -45,89 +47,161 @@
+        while( ! ready );
+    }
+
+
 protected:
-   /**
-    * initialize - internal method to be used within the run method
-    * takes care of the initialization using the already allocated
-    * FIFO object passed as a param.  This function will throw 
-    * an exception if either port (src or dst) have already been
-    * allocated.
-    * @param   src - PortInfo*, nullptr if not to be set
-    * @param   dst - PortInfo*, nullptr if not to be set
-    * @param   fifo - FIFO*
-    * @throws  PortDoubleInitializeException - if either port is already initialized.
-    */
-   void initialize( PortInfo * const src, 
-                    PortInfo * const dst, 
-                    FIFO * const fifo );
-
-   
-   virtual void allocate( PortInfo &a, PortInfo &b, void *data );
-
-   /**
-    * setReady - call within the implemented run function to signal
-    * that the initial allocations have been completed.
-    */
-   void setReady() ;
-
-   /** both convenience structs, hold exactly what the names say **/
-   kernelkeeper   &source_kernels;
-   kernelkeeper   &all_kernels;
-   
-   /** 
-    * keeps a list of all currently allocated FIFO objects,
-    * set from within the initialize function.
-    */
-   std::set< FIFO*   > allocated_fifo;
-
-   /**
-    * exit_alloc - bool whose value is set by the map 
-    * object, controls when the loop within the alloc thread
-    * is exited.
-    */
-   volatile bool &exit_alloc;
+    /**
+     * initialize - internal method to be used within the run method
+     * takes care of the initialization using the already allocated
+     * FIFO object passed as a param.  This function will throw
+     * an exception if either port (src or dst) have already been
+     * allocated.
+     * @param   src - PortInfo*, nullptr if not to be set
+     * @param   dst - PortInfo*, nullptr if not to be set
+     * @param   fifo - FIFO*
+     * @throws  PortDoubleInitializeException -
+     *          if either port is already initialized.
+     */
+    void initialize( PortInfo * const src,
+                     PortInfo * const dst,
+                     FIFO * const fifo )
+    {
+        assert( fifo != nullptr );
+        assert( dst != nullptr );
+        assert( src != nullptr );
+        if( src->getFIFO() != nullptr )
+        {
+            throw PortDoubleInitializeException(
+                    "Source port \"" +
+                    src->my_kernel->output.getPortName( src->my_name ) +
+                    "\" already initialized!" );
+        }
+        if( dst->getFIFO() !=  nullptr )
+        {
+            throw PortDoubleInitializeException(
+                    "Destination port \"" +
+                    dst->my_kernel->input.getPortName( dst->my_name ) +
+                    "\" already initialized!" );
+        }
+        src->setFIFO( fifo );
+        dst->setFIFO( fifo );
+        /** NOTE: this list simply speeds up the monitoring if we want it **/
+        allocated_fifo.insert( fifo );
```

Comment:
Need to double check that the monitoring thread (if it's used) still traverses this structure (e.g., to dump periodic frames of queue stats.